geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ClientHealthMonitor.java

```diff
@@ -351,18 +333,19 @@ public void removeConnection(ClientProxyMembershipID proxyID, ServerConnection c
    * @param proxyID The id of the client from which the ping was received
    */
   public void receivedPing(ClientProxyMembershipID proxyID) {
-    if (this._clientMonitor == null) {
+    if (_clientMonitor == null) {
       return;
     }
+
     if (logger.isTraceEnabled()) {
       logger.trace("ClientHealthMonitor: Received ping from client with member id {}", proxyID);
     }
-    synchronized (_clientHeartbeatsLock) {
-      if (!this._clientHeartbeats.containsKey(proxyID)) {
-        registerClient(proxyID);
-      } else {
-        this._clientHeartbeats.put(proxyID, Long.valueOf(System.currentTimeMillis()));
-      }
+
+    AtomicLong heartbeat = _clientHeartbeats.get(proxyID);
```

Comment:
It seems like this logic should all be in registerClient and this code should just call registerClient. I don't understand why the old code split this code up into two places.

Reply:
I recall there being some other code paths that came in where it made sense to split them. Something about registering without updating the time or something like that. Basically I tried to leave the logic the same and just pull out the locking. Wasn't looking to refactor the insanity in here.

Reply:
But this code in receivePing is basically just duplicating logic in registerClient. If it finds an entry in clientHeartbeats it updates it by setting it to currentTimeMillis. If it doesn't find one then it calls registerClient which will then just do all that again. I don't see any value in having this logic duplicated. The only difference is we call "get" here first (which is what I suggested we call first in registerClient).

Reply:
Ok, I see it now. When I was whittling down the old methods I didn't see this optimization. Thanks!

Reply:
Actually it turns out the original didn't have an update time in register if it was already registered. Added tests to constrain the refactor and cleaned this up a bit more.